src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/VsTelemetryService.cs

```diff
@@ -39,5 +161,88 @@ public TelemetryEventCorrelation Report(string eventName, string description, Ex
+        /// <returns>Hashed value.</returns>
+        public string HashValue(string value)
+        {
+            using (var crypto = SHA256.Create())
+            {
+                byte[] hash = crypto.ComputeHash(Encoding.UTF8.GetBytes(value));
+                return BitConverter.ToString(hash);
+            }
+        }
+
+
+        /// <summary>
+        /// Build a fully qualified property name based on it's parent event and the property name
+        /// </summary>
+        /// <param name="eventName">Name of the parent event.</param>
+        /// <param name="propertyName">Name of the property.</param>
+        /// <returns>Fully qualified property name.</returns>
+        /// <remarks>
+        /// Properties are expected to be in the following format - EventName.PropertyName
+        /// with the the slashes from the vent name replaced by periods.
+        /// e.g. vs/myevent would translate to VS.MyEvent.MyProperty
+        /// </remarks>
+        private string BuildPropertyName(string eventName, string propertyName)
+        {
+            string name = PropertyPrefix + eventName.Replace('/', '.');
+
+            if (!name.EndsWith("."))
+            {
+                name += ".";
+            }
+
+            name += propertyName;
+
+            return name;
+        }
+
+        /// <summary>
+        /// Tries to correlate a project 
+        /// </summary>
+        /// <param name="telemetryEvent">Telemetry event to correlate the asset to.</param>
+        private void PostEvent(TelemetryEvent telemetryEvent)
+        {
+            if (!_telemetryEventCorrelationInitialized)
+            {
+                _unconfiguredProjectCommonServices.ThreadingService.Fork(async () =>
+                {
+                    if (await CreateProjectCorrelationAssetAsync().ConfigureAwait(false))
+                    {
+                        PostEvent(telemetryEvent, true);
+                    }
+                    else
+                    {
+                        // Post the event even if correlation asset cannot be created.
+                        // This will let us know about the existence of cases where asset creation falied after project initialization.
+                        PostEvent(telemetryEvent, false);
+                    }
+                }, unconfiguredProject: _unconfiguredProjectCommonServices.Project);
+            }
+            else
+            {
+                PostEvent(telemetryEvent, true);
+            }
+        }
+
+        /// <summary>
+        /// Tries to correlate a project 
+        /// </summary>
+        /// <param name="telemetryEvent">Telemetry event to correlate the asset to.</param>
+        /// <param name="correlate">Says whether the telemetry event should be correlated</param>
+        private void PostEvent(TelemetryEvent telemetryEvent, bool correlate)
+        {
+            if (correlate)
+            {
+                telemetryEvent.Correlate(_telemetryEventCorrelation);
```

Comment:
What does the Correlate call do to a telemetryevent? Does this just add the correlation ID to the event?